sphinx/util/inspect.py

```diff
@@ -241,14 +239,62 @@ def isclassmethod(
     return False
 
 
+def is_classmethod_descriptor(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> TypeIs[types.ClassMethodDescriptorType]:
+    """Check if the object is a :class:`~types.ClassMethodDescriptorType`.
+
+    This is typically useful to check if a built-in method is a class method.
+    """
+    if isinstance(obj, types.ClassMethodDescriptorType):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return isinstance(meth, types.ClassMethodDescriptorType)
+    return False
+
+
+def is_builtin_classmethod(obj: Any, cls: Any = None, name: str | None = None) -> bool:
+    """Check if the object is a class method implemented in C."""
+    if is_classmethod_descriptor(obj, cls, name):
+        return True
+    if (
+        isbuiltin(obj)
+        and getattr(obj, '__self__', None) is not None
+        and isclass(obj.__self__)
+    ):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return is_classmethod_descriptor(obj, meth, name) or (
+                    isbuiltin(obj)
+                    and getattr(obj, '__self__', None) is not None
+                    and isclass(obj.__self__)
+                )
```

Comment:
Previously we had `is_builtin_classmethod(meth)`, so `cls=None` and `name=None`, so the inlined code would actually be:

```suggestion
                return is_classmethod_descriptor(meth, None, None) or (
                    isbuiltin(meth)
                    and getattr(meth, '__self__', None) is not None
                    and isclass(meth.__self__)
                )
```

Sorry for the error. However it seems that there is no test coverage of this as tests passed?

Reply:
I think we exercise this code path when `obj` is a junk object but we know the base class and the name (e.g., `obj = None` because we tried to retrieve on the `__dict__` of a child class). 

I can try to write a test by the end of the week (I'm leaving tomorrow for a few days; if I manage to write something tomorrow, I'll add a commit, otherwise I'll only be able to write something on Friday or later)